File: atlases/relatedness/shared/loaders/inversion_karyotypes.js
// shared/loaders/inversion_karyotypes.js
// =============================================================================
// Loads the per-sample × per-candidate karyotype matrix from the Inversion
// Atlas catalogue export. Expected format: long-format TSV with one row per
// (sample_id, inversion_id) plus a karyotype column ('0/0' | '0/1' | '1/1'
// | 'NA') and an optional quality column ('high' | 'low').
//
// Drives the Karyotypes page (pages/hub/karyotypes.js), the Mendelian
// tester (pages/hub/mendelian.js), and the per-row family roster on the
// Inversions page (pages/hub/inversions.js).
//
// Extract-schema:
//   {
//     "schema":             "inversion_karyotypes_v1",
//     "produced_by":        "Inversion Atlas catalogue export",
//     "samples":            string[],
//     "inversions":         Array<{ candidate, chromosome, start_mb, end_mb,
//                                    length_mb, frequency, status, notes }>,
//     "karyotype_matrix":   Object<sample_id → Object<inv_id → karyotype>>,
//     "karyotype_quality":  Object<sample_id → Object<inv_id → 'high'|'low'>>,
//   }
// =============================================================================

import { readTsv } from '../api_client.js';

export async function loadInversionKaryotypes(args, { fetcher = null } = {}) {
  const { karyotype_tsv_path, inversion_catalogue_path } = args || {};
  if (!karyotype_tsv_path) {
    throw new Error('loadInversionKaryotypes: karyotype_tsv_path is required');
  }

  const ktData = fetcher
    ? fetcher(karyotype_tsv_path)
    : await readTsv(karyotype_tsv_path);
  const ktRows = ktData.rows;

  const samples_set = new Set();
  const inv_set     = new Set();
  const karyotype_matrix  = {};
  const karyotype_quality = {};
  for (const r of ktRows) {
    const sid = r.sample_id || r.sample;
    const iid = r.candidate || r.inversion_id || r.inv_id;
    if (!sid || !iid) continue;
    samples_set.add(sid);
    inv_set.add(iid);
    if (!karyotype_matrix[sid])  karyotype_matrix[sid] = {};
    if (!karyotype_quality[sid]) karyotype_quality[sid] = {};
    karyotype_matrix[sid][iid]  = r.karyotype || 'NA';
    karyotype_quality[sid][iid] = r.quality   || 'high';
  }

  // Optional second-file: the candidate catalogue from the Inversion Atlas.
  let inversions = [];
  if (inversion_catalogue_path) {
    const cat = fetcher
      ? fetcher(inversion_catalogue_path)
      : await readTsv(inversion_catalogue_path);
    inversions = cat.rows.map(r => ({
      candidate:   r.candidate || r.inversion_id,
      chromosome:  r.chromosome || r.chrom,
      start_mb:    parseFloat(r.start_mb),
      end_mb:      parseFloat(r.end_mb),
      length_mb:   parseFloat(r.length_mb || (parseFloat(r.end_mb) - parseFloat(r.start_mb))),
      frequency:   parseFloat(r.frequency || '0'),
      status:      r.status || 'pass',
      notes:       r.notes || '',
    }));
  } else {
    // Fall back to the inv ids found in the karyotype matrix, with
    // placeholder coordinates. Pages can render the matrix but the
    // Inversions table will be sparse.
    inversions = Array.from(inv_set).sort().map(c => ({
      candidate:   c,
      chromosome:  '?',
      start_mb:    NaN, end_mb: NaN, length_mb: NaN,
      frequency:   NaN,
      status:      'unknown',
      notes:       '',
    }));
  }

  return {
    schema:            'inversion_karyotypes_v1',
    produced_by:       'Inversion Atlas catalogue export',
    produced_at:       new Date().toISOString(),
    samples:           Array.from(samples_set),
    inversions,
    karyotype_matrix,
    karyotype_quality,
  };
}

File: atlases/relatedness/registries/extractors/inversion_karyotypes.py
"""extractor: inversion_karyotypes_v1 — per-sample × per-inversion karyotype.

Long-format TSV from the Inversion Atlas catalogue export. Drives the
Karyotypes tab, the Mendelian tester, and the four-stage scoring on
Inversions.

Expected columns: sample_id, inversion_id, karyotype ('0/0'|'0/1'|'1/1'|'NA'),
optional quality.

Emit:
    { records: [...], n_records, n_samples, n_inversions,
      karyotype_counts: { '0/0', '0/1', '1/1', 'NA' },
      _provenance: {...} }
"""
from __future__ import annotations
import collections
import pathlib
from typing import Any, Dict
from . import _tsv as _t


def extract(raw_outputs: Dict[str, str], params: Dict[str, Any]) -> Dict[str, Any]:
    path = pathlib.Path(raw_outputs["file_path"])
    _, rows = _t.read_tsv(path, has_header=True, infer_types=False,
                          max_rows=int(params.get("max_rows") or 0))

    samples = set()
    inversions = set()
    karyo_counts: Dict[str, int] = collections.Counter()
    records = []
    for r in rows:
        sid = r.get("sample_id") or r.get("sample") or r.get("id")
        inv = r.get("inversion_id") or r.get("candidate_id") or r.get("inv_id")
        if sid is None or inv is None:
            continue
        k = r.get("karyotype") or r.get("genotype")
        out = {"sample_id": str(sid), "inversion_id": str(inv), "karyotype": k}
        if "quality" in r: out["quality"] = r["quality"]
        records.append(out)
        samples.add(str(sid))
        inversions.add(str(inv))
        karyo_counts[str(k) if k is not None else "NA"] += 1

    return {
        "records":           records,
        "n_records":         len(records),
        "n_samples":         len(samples),
        "n_inversions":      len(inversions),
        "karyotype_counts":  dict(karyo_counts),
        "_provenance":       _t.provenance(raw_outputs, path, row_count=len(records)),
    }
